---
title: "Sample Failure Analysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Summary
Review read count for samples that failed within project.
```{r echo=FALSE,include=FALSE}
project_id="OH-M5185-220406"
project_date="220406"

```


```{r echo=FALSE,include=FALSE}
library(ggplot2)
```


```{r echo=FALSE,include=FALSE}
l_drive=("L:Micro\\Gen Micro\\Whole Genome Sequencing\\Coronavirus_WGS\\COVID-19 Fastas\\")

#pull analysis_file
df_analysis=read.csv(paste0(l_drive,project_id,"\\analysis\\final_results_20",project_date,".csv"))

#pull cecret_file
df_cecret=read.csv(paste0(l_drive,project_id,"\\analysis\\intermed\\cecret_summary.csv"))
head(df_cecret)

#merge data
df_merged=merge.data.frame(df_analysis,df_cecret)

# cleanup sample name
df_merged$sample_id = gsub(paste0("-",project_id), "", df_merged$sample_id)
```

```{r echo=FALSE,include=FALSE}
fq1_fail = sapply(df_merged[1:2], function(x) mean(df_merged$fastqc_raw_reads_1[x == "fail"]))[[2]]
fq1_pass = sapply(df_merged[1:2], function(x) mean(df_merged$fastqc_raw_reads_1[x == "passed_qc"]))[[2]]
fq2_fail = sapply(df_merged[1:2], function(x) mean(df_merged$fastqc_raw_reads_2[x == "fail"]))[[2]]
fq2_pass = sapply(df_merged[1:2], function(x) mean(df_merged$fastqc_raw_reads_2[x == "passed_qc"]))[[2]]
seq_fail = sapply(df_merged[1:2], function(x) mean(df_merged$seqyclean_pairs_kept_after_cleaning[x == "fail"]))[[2]]
seq_pass = sapply(df_merged[1:2], function(x) mean(df_merged$seqyclean_pairs_kept_after_cleaning[x == "passed_qc"]))[[2]]

df_merged$avfq1=0
df_merged$avfq2=0
df_merged$avseq=0

df_merged <- within(df_merged, avfq1[pango_qc == 'fail'] <- fq1_fail)
df_merged <- within(df_merged, avfq1[pango_qc == 'passed_qc'] <- fq1_pass)
df_merged <- within(df_merged, avfq2[pango_qc == 'fail'] <- fq2_fail)
df_merged <- within(df_merged, avfq2[pango_qc == 'passed_qc'] <- fq2_pass)
df_merged <- within(df_merged, avseq[pango_qc == 'fail'] <- seq_fail)
df_merged <- within(df_merged, avseq[pango_qc == 'passed_qc'] <- seq_pass)


```

### Raw Data
Review raw pairs by PANGOLIN QC

```{r echo=FALSE}
ggplot(data=df_merged,aes(x=sample_id,y=fastqc_raw_reads_1 )) +
  facet_wrap(~pango_qc,scales = "free_x") +
  geom_bar(stat="identity", position=position_dodge()) +
  ggtitle("Raw FASTQ Read Count Read #1") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_hline(aes(yintercept = avfq1, group = sample_id), colour = 'red') +
  geom_text(aes(4,avfq1,label = format(round(avfq1,2),nsmall=2), vjust = -5),colour = 'red')


ggplot(data=df_merged,aes(x=sample,y=fastqc_raw_reads_2 )) +
  facet_wrap(~pango_qc,scales = "free_x") +
  geom_bar(stat="identity", position=position_dodge()) +
  ggtitle("Raw FASTQ Read Count Read #2") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_hline(aes(yintercept = avfq2, group = sample_id), colour = 'red') +
  geom_text(aes(4,avfq2,label = format(round(avfq2,2),nsmall=2), vjust = -5),colour = 'red')

```

### Filtered Data
Review filtered read pairs by PANGOLIN QC

```{r echo=FALSE}
ggplot(data=df_merged,aes(x=sample,y=seqyclean_pairs_kept_after_cleaning)) +
  facet_wrap(~pango_qc,scales = "free_x") +
  geom_bar(stat="identity", position=position_dodge()) +
  ggtitle("Filtered FASTQ Read Count for Pairs") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_hline(aes(yintercept = avseq, group = sample_id), colour = 'red') +
  geom_text(aes(4,avseq,label = format(round(avseq,2),nsmall=2), vjust = -5),colour = 'red')
```

