---
title: "LiquidHandler"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Background

Assess the quality of samples using two different liquid handlers: epMotion (standard) vs Hamilton. Fragments are first reviewed, which include adaptor and read sequences. Reads are then assessed once adaptors have been trimmed. Finally lineage results were reviewed.

```{r echo=FALSE, warning=FALSE}
library(ggplot2)
library(ggrepel)

```

```{r echo=FALSE}
# read in files
df_frag_raw=read.csv("merged_frag.csv")
colnames(df_frag_raw)=c("len","count","id","liquid_handler")

df_stats_raw=read.csv("merged_coverage.csv")
colnames(df_stats_raw)=c("id","liquid_handler","uniquely_mapped_reads","cov_ov_100",
                     "cov_50_100","cov_20_50","cov_15_20",  "cov_10_15","cov_3_10",
                     "cov_1_3","cov_0_1","input_reads","dup_reads","total_unique_reads","qc_reads_fail",
                     "total_mapped_reads","map_perc","unmapped_reads","mapq_40","mapq_30_40","mapq_20_30",
                     "mapq_10_20","mapq_0_10")
#align_reads is number of unique reads that mapped
#uni_reads is the number of unique reads whether or not they mapped
#map_count is total number of reads that mapped
#input_reads is number of mapped or unmapped reads

df_lin_raw=read.csv("final_results_20220425.csv")
colnames(df_lin_raw)=c("id","pango_qc","nextclade_clade","pangolin_lineage","pangolin_scorpio","aa_substitutions")
```

```{r echo=FALSE}
# subset to only include samples with both information
sample_list=unique(subset(df_frag_raw,liquid_handler=="Hamilton")$id)
sample_list=append(sample_list,gsub("b","",sample_list))

df_frag=subset(df_frag_raw,id %in% sample_list)
df_stats=subset(df_stats_raw,id %in% sample_list)
df_lin=subset(df_lin_raw,id %in% sample_list)

remove(df_frag_raw,df_stats_raw)

df_stats$sample=gsub("b","",df_stats$id)
df_lin$sample=gsub("b","",df_lin$id)
```

### Review sample information

```{r echo=FALSE}
print(paste0("Total Samples: ",length(unique(df_frag$id))))
print(paste0("Total epMotion: ",length(unique(subset(df_frag,liquid_handler=="epMotion")$id))))
print(paste0("Total Hamilton: ",length(unique(subset(df_frag,liquid_handler=="Hamilton")$id))))
```

### Review fragment size by liquid handler at project level

- To determine read pairs, Illumina employs a Gaussian statistics model to evaluate the likelihood of two reads being matched. Within this modeling, fragment distribution is calculated and resulted in an output file. This file was reviewed. Each sample has approxmiately 100,000 fragments sub-sampled to represent the overall quality of the sample.

- The overall fragment distribution appears to vary by liquid handler type. Due to the left-sided density, however, this is difficult to interpret raw data.

```{r echo=FALSE}
ggplot(data=df_frag,aes(x=len,y=count))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler) +
  ggtitle("Frragment length by count for each liuqid handler")

```

- To improve visualization, fragments were binned into groups by length. After binning some variation is noted between the two liquid handlers, however, both had the majority of fragments within the expected 100-500bp length bin (epMotion 54%, Hamilton 51%).

```{r echo=FALSE}
#create bins for frag
dat=data.frame()
for (sample_id in unique(df_frag$id)){
  sub_df=subset(df_frag,id==sample_id)

  # set up cut-off values 
  breaks <- c(0,100,500,1000,5000,10000,15000,20000,25000,30000)
  
  # specify interval/bin labels
  tags <- c("[0-100)","[100-500)", "[500-1000)", "[1000-5000)", "[5000-10000)",
            "[1000-15000)","[15000-20000)", "[20000-25000)","[25000-30000)")
    
  # bucketing values into bins
  sub_df$tags <- cut(sub_df$len, 
                      breaks=breaks, 
                      include.lowest=TRUE, 
                      right=FALSE, 
                      labels=tags)
  dat=rbind(dat,sub_df)
}

# collapse to liquid handler
df_frag_sub=data.frame()
for (liq_id in unique(dat$liquid_handler)){
  sub_df1=subset(dat,(liquid_handler==liq_id))
  
  for (tag_id in unique(dat$tags)){
    sub_df2=subset(sub_df1,tags==tag_id)

    df_frag_sub[nrow(df_frag_sub)+1,"liquid_handler"]=liq_id
    df_frag_sub[nrow(df_frag_sub),"bin"]=tag_id
    df_frag_sub[nrow(df_frag_sub),"perc"]=round((sum(sub_df2$count)/sum(sub_df1$count))*100,2)
  }
}
df_frag_sub=subset(df_frag_sub,perc>0)
ggplot(data=df_frag_sub,aes(x=bin,y=perc,fill=bin))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler) +
  ggtitle("Frragment length binned, for each liquid handler")+
  xlab('bin length') + ylab('Percent of fragments') +
  geom_text(aes(label=perc), position=position_dodge(width=0.9), vjust=-0.25)+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

### Review fragment size by liquid handler at sample level

- At a sample level, some variation is observed, but these values do not exceed what we might expect from laboratory processing between runs.

```{r echo=FALSE}
#merge counts by sample
df_frag_sub=data.frame()
for (sample_id in unique(dat$id)){
  sub_df=subset(dat,id==sample_id)
  cumulative_count=0
  
  for (tagid in sort(unique(sub_df$tags))){
    tag_count=subset(sub_df,tags==tagid)$count
    total_count=sub_df$count
    cumulative_count=sum(tag_count)+cumulative_count

    df_frag_sub[nrow(df_frag_sub)+1,"tags"]=tagid
    df_frag_sub[nrow(df_frag_sub),"id"]=gsub("b","",sample_id)
    df_frag_sub[nrow(df_frag_sub),"liquid_handler"]=unique(sub_df$liquid_handler)
    df_frag_sub[nrow(df_frag_sub),"counts"]=sum(subset(sub_df,tags==tagid)$count)
    df_frag_sub[nrow(df_frag_sub),"percent"]=paste0(round(sum(tag_count)/sum(total_count)*100,1),"%")
    df_frag_sub[nrow(df_frag_sub),"cum"]=cumulative_count
  }
}
df_frag_sub=subset(df_frag_sub,percent!="0%")
remove(dat)

for (i in seq(1,length(unique(df_frag_sub$id)),8)){
  sample_list=unique(df_frag_sub$id)[i:(i+10)]
  sub_df=subset(df_frag_sub,id %in% sample_list)
  p = ggplot(data=sub_df, aes(x=liquid_handler, y=counts, fill=tags)) +
    geom_bar(stat="identity") +
    geom_text_repel(aes(y=cum,label=percent),direction="y",box.padding=0) +
    facet_wrap(~id, scales="free")+
    xlab('Liquid Handler') + ylab('Number of Fragments') +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
  print(p)
}
```

### Review input reads

- Reviewing all usable reads for analysis indicates stark differences in the number of reads between the liquid handlers. The total count of reads is significantly higher in the Hamilton samples over the epMotion samples.This is consistent with the % Reads Identified (PF) Per Index report generated via DRAGEN

```{r echo=FALSE}
ggplot(data=df_stats,aes(x=liquid_handler,y=input_reads,fill=liquid_handler))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler,scales="free") +
  ggtitle("Frragment length binned, for each liquid handler")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

### Review reads by duplication type

- Focusing on read quality, reads were classified as unique or duplicate. Both in raw values, and percentage, duplicated reads make up a significant portion of the Hamilton reads. This indicates that the same read was being amplified and sequenced repeatedly in this sample.

```{r echo=FALSE}
df_stat_sub=data.frame()
for (liq_id in unique(df_stats$liquid_handler)){
  sub_df=subset(df_stats,liquid_handler==liq_id)
  
  df_stat_sub[nrow(df_stat_sub)+1,"liquid_handler"]=liq_id
  df_stat_sub[nrow(df_stat_sub),"reads"]=sum(sub_df$total_unique_reads)
  df_stat_sub[nrow(df_stat_sub),"type"]="unique"
  df_stat_sub[nrow(df_stat_sub),"cum"]=sum(sub_df$total_unique_reads)
  df_stat_sub[nrow(df_stat_sub),"perc"]=sum(sub_df$total_unique_reads)/(sum(sub_df$total_unique_reads)+sum(sub_df$dup_reads))*100

  
  df_stat_sub[nrow(df_stat_sub)+1,"liquid_handler"]=liq_id
  df_stat_sub[nrow(df_stat_sub),"reads"]=sum(sub_df$dup_reads)
  df_stat_sub[nrow(df_stat_sub),"type"]="duplicated"
    df_stat_sub[nrow(df_stat_sub),"cum"]=sum(sub_df$dup_reads)
  df_stat_sub[nrow(df_stat_sub),"perc"]=sum(sub_df$dup_reads)/(sum(sub_df$total_unique_reads)+sum(sub_df$dup_reads))*100
}

ggplot(data=df_stat_sub,aes(x=liquid_handler,y=reads ,fill=type))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler, scales="free") +
  ggtitle("Total reads, by type, for each liuqid handler")+
  geom_text_repel(aes(y=cum,label=round(perc,2)),direction="y",box.padding=0)

```


### Review aligned reads by liquid handler

- Next, reads were reviewed for mapping.The percentage of mapped to unmapped reads (98.73 to 98.81) is not significantly different

```{r echo=FALSE}
df_stat_sub=data.frame()
for (liq_id in unique(df_stats$liquid_handler)){
  sub_df=subset(df_stats,liquid_handler==liq_id)
  
  df_stat_sub[nrow(df_stat_sub)+1,"liquid_handler"]=liq_id
  df_stat_sub[nrow(df_stat_sub),"reads"]=sum(sub_df$unmapped_reads)
  df_stat_sub[nrow(df_stat_sub),"type"]="unmapped"
  df_stat_sub[nrow(df_stat_sub),"cum"]=sum(sub_df$unmapped_reads)
  df_stat_sub[nrow(df_stat_sub),"perc"]=sum(sub_df$unmapped_reads)/(sum(sub_df$unmapped_reads)+sum(sub_df$total_mapped_reads))*100

  
  df_stat_sub[nrow(df_stat_sub)+1,"liquid_handler"]=liq_id
  df_stat_sub[nrow(df_stat_sub),"reads"]=sum(sub_df$total_mapped_reads)
  df_stat_sub[nrow(df_stat_sub),"type"]="mapped"
    df_stat_sub[nrow(df_stat_sub),"cum"]=sum(sub_df$total_mapped_reads)
  df_stat_sub[nrow(df_stat_sub),"perc"]=sum(sub_df$total_mapped_reads)/(sum(sub_df$unmapped_reads)+sum(sub_df$total_mapped_reads))*100
}

ggplot(data=df_stat_sub,aes(x=liquid_handler,y=reads ,fill=type))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler, scales="free") +
  ggtitle("Aligned reads each liuqid handler")+
  geom_text_repel(aes(y=cum,label=round(perc,2)),direction="y",box.padding=0)
```

- Coverage was then reviewed. First, areas with at least 100X coverage was reviewed, with an average across all samples of the epMotion (80.47) higher than the Hamilton (76.69). One sample (2022047652) is found to have a lower fraction of reads with 100X coverage on both liquid handlers. This means while there are more reads resulting from the Hamilton, it is at a lower depth of coverage overall.

```{r echo=FALSE}
ggplot(data=df_stats,aes(x=id,y=cov_ov_100,fill=liquid_handler))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler, scales="free") +
  ggtitle("Percent coverage at least 100X")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_text_repel(aes(y=cov_ov_100,label=round(cov_ov_100,2)),direction="y",box.padding=0)

```

- Next, reviewing all samples at the lowest coverage (0X to 1X), we see that for most samples, the Hamilton produces a greater percentage of the genome with 0-1X coverage. On average, this is a nearly 5% difference between the Hamilton and epMotion (21.69% as compared to 17.35%, respectively). 

```{r echo=FALSE}
cov_list=c("cov_50_100","cov_20_50","cov_15_20","cov_10_15","cov_3_10","cov_1_3","cov_0_1")
df_stats_cov=data.frame()
for (cov_id in cov_list){
  for (liq_id in unique(df_stats$liquid_handler)){
    df_stats_cov[nrow(df_stats_cov)+1,"id"]=cov_id
    df_stats_cov[nrow(df_stats_cov),"value"]=mean(subset(df_stats,liquid_handler==liq_id)[,cov_id])
    df_stats_cov[nrow(df_stats_cov),"liquid_handler"]=liq_id
  }
}
ggplot(data=df_stats_cov,aes(x=liquid_handler,y=value,fill=liquid_handler)) +
  geom_bar(stat="identity") + 
  facet_wrap(~id) +
  ggtitle( "Percent coverage from 0 through 100X") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_text_repel(aes(y=value,label=value),direction="y",box.padding=0)

```

- Finally, MAPQ values were reviewed. There was no signfiicant difference between the percent of reads with a MAPQ >40 between the epMotion and Hamilton samples (99.8% and 99.78% respectively).

```{r echo=FALSE}
mapq_list=c("mapq_40","mapq_30_40","mapq_20_30","mapq_10_20","mapq_0_10")
df_stats_mq=data.frame()
for (map_id in mapq_list){
  for (liq_id in unique(df_stats$liquid_handler)){
    df_stats_mq[nrow(df_stats_mq)+1,"id"]=map_id
    df_stats_mq[nrow(df_stats_mq),"value"]=mean(subset(df_stats,liquid_handler==liq_id)[,map_id])
    df_stats_mq[nrow(df_stats_mq),"perc"]=(mean(subset(df_stats,liquid_handler==liq_id)[,map_id])/mean(subset(df_stats,liquid_handler==liq_id)$total_mapped_reads))*100
    df_stats_mq[nrow(df_stats_mq),"liquid_handler"]=liq_id
  }
}
ggplot(data=df_stats_mq,aes(x=id,y=perc,fill=id))+
  geom_bar(stat="identity")+
  facet_grid(~liquid_handler, scales="free") +
  ggtitle("MAPQ Scores by liquid handler")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_text_repel(aes(y=perc,label=round(perc,2)),direction="y",box.padding=0)

```

### Review analysis results

- Ultimately, while the above results show some differences, none are alone alarming. The results from these statistics must then be used to determine if further investigation is warranted. Consensus was found between all samples for nextclade and pangolin lineage/clade calls. Discordance was observed, however, within the aa_substitions for several samples.

```{r echo=FALSE}
check_list=c("nextclade_clade","pangolin_lineage","pangolin_scorpio","aa_substitutions")
for (check_id in check_list){
  print (paste0("** CHECKING: ",check_id,"**"))
  for (sample_id in unique(df_lin$sample)){
    ep_sample=sample_id
    ham_sample=paste0(sample_id,"b")
    
    ep_sub=subset(df_lin,id==ep_sample)
    ham_sub=subset(df_lin,id==ham_sample)
    
    if(ep_sub[,check_id]!=ham_sub[,check_id]){
      print(paste0("----Discordant for sample ",sample_id))
    } else{
      print(paste0("----Consensus for sample ",sample_id))
    }
  }
}
```

- To determine the extent of discordance, each mutation was reviewed and unique lists analyzed. No mutations were only found in epMotion samples, however, additional mutations were found in Hamilton samples.

```{r echo=FALSE}

check_id="aa_substitutions"
for (sample_id in unique(df_lin$sample)){
  ep_sample=sample_id
  ham_sample=paste0(sample_id,"b")
    
  ep_sub=subset(df_lin,id==ep_sample)
  ham_sub=subset(df_lin,id==ham_sample)
  
  if(ep_sub[,check_id]!=ham_sub[,check_id]){
    print(paste0("--Discordant analysis for ", sample_id,""))
    ep_list=strsplit(ep_sub[,check_id],",")[[1]]
    ham_list=strsplit(ham_sub[,check_id],",")[[1]]
    
    missing_list1=as.list(ep_list[!(ep_list %in% ham_list)])
    missing_list2=as.list(ham_list[!(ham_list %in% ep_list)])
    #print("--Mutations found in epMotion samples only: ")
    #print(paste0("----",missing_list1))
    print("----Mutations found in Hamilton samples only: ")
    print(paste0("------",missing_list2))
  }
}
```


