---
title: "MutationAnalysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

## User Input
```{r echo=FALSE}
#set projects
project_list=c("OH-VH00648-220106", "OH-M2941-220111", "OH-VH00648-211221", "OH-M2941-211222", 
               "OH-M6588-220125", "OH-VH00648-220126", "OH-VH00648-220201", "OH-VH00648-220215",
               "OH-M5185-220406 SARS only")

mutation_list=c("S:L452Q","S:L452R")
```

```{r echo=FALSE}
print("These are the included projects: ")
print(project_list)

print("These are the included mutations: ")
print(mutation_list)
```


```{r echo=FALSE,include=FALSE, warning=FALSE}
#set dir
input_dir=("L:Micro\\Gen Micro\\Whole Genome Sequencing\\Coronavirus_WGS\\COVID-19 Fastas\\")
output_dir=(paste0(input_dir,"mutation_analysis\\"))

#load libraries
library(ggplot2)
library(knitr)
```

```{r echo=FALSE,include=FALSE}
mutation_df=data.frame()
for (project_id in project_list){
  # pull date info from project_id, add "20" to include year
  project_id=paste0("",gsub(" [A-Za-z]*","",project_id))
  date_stamp=paste0(20,gsub("OH-VH[0-9]*-","",project_id))
  date_stamp=paste0("",gsub("OH-M[0-9]*-","",date_stamp))
  
  # read in analysis file
  analysis_raw=read.csv(paste0(input_dir,project_id,"\\analysis\\final_results_",date_stamp,".csv"),
                       na.strings=c("","NA"))
  
  # subset for those that passed pango_QC
  analysis_df=subset(analysis_raw,pangolin_qc=="passed_qc")
  
  # pull all mutations for each sample
  mutation_array=strsplit(analysis_df$aa_substitutions,",")
  
  # for each sample, review the mutations found
  # pull unique mutation list for each sample
  # if a mutation is found in df, add to count
  # if it is  not found then create a new row with count of 1
  mutation_tmp=data.frame()
  for (i in 1:length(mutation_array)){
    mutation_row=paste0(analysis_df[i,"pangolin_lineage"],"+",mutation_array[[i]],"+",date_stamp)
    mutation_row=unique(mutation_row)
    
    for (mutation_id in mutation_row){
      if (is.na(mutation_id)){
        next
      } else if (any(row.names(mutation_tmp) == mutation_id)){
        mutation_tmp[mutation_id,"value"]=mutation_tmp[mutation_id,"value"]+1
        mutation_tmp[mutation_id,"mutation"]=strsplit(mutation_id,"[+]")[[1]][2]
        mutation_tmp[mutation_id,"lineage"]=analysis_df[i,"pangolin_lineage"]
        mutation_tmp[mutation_id,"date_stamp"]=date_stamp
       } else{
         mutation_tmp[mutation_id,"value"]=1
         mutation_tmp[mutation_id,"mutation"]=strsplit(mutation_id,"[+]")[[1]][2]
         mutation_tmp[mutation_id,"lineage"]=analysis_df[i,"pangolin_lineage"]
         mutation_tmp[mutation_id,"date_stamp"]=date_stamp
     }
    }
  }
  
  # determine number of total mutations
  analysis_df$total_mut=count.fields(textConnection(analysis_df$aa_substitutions), sep = ",")
  
  # determine percentage of samples with mutation
  mutation_tmp$percent_mutation=round((mutation_tmp$value/nrow(analysis_df))*100,2)
  
  #merge data
  #analysis_merge=rbind(analysis_merge,analysis_df)
  mutation_df=rbind(mutation_df,mutation_tmp)
}
```

## Results
### Review top mutations per time_point
"Highest Rate" was determined by calculating the percent of a particular mutation being found in across all samples at each time point. Each lineage within a time point was reviewed, and the highest mutation value was plotted. For example:

```{r echo=FALSE}
tmp_df=data.frame()
kable(mutation_df[c(1,5,7,200),c("mutation","lineage","percent_mutation")])
```

The highest mutation for BA.1.1 is 1.33% and for AY.25 is 10.61%. Mutations matching this value per lineage, are plotted. Note: If more than 6 lineages are present within a time point, multiple graphs are created to ease viewing.

```{r echo=FALSE, out.height="150%"}
plot_by_date<-function(df_in){
      p = ggplot(data=df_in,aes(x=mutation,y=percent_mutation )) +
        facet_wrap(~lineage,scales = "free_x",ncol=3) + 
        geom_bar(stat="identity", position=position_dodge()) +
        ggtitle(paste0("Percent of Samples with mutation by timepoint, by lineage"),date_id) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text = element_text(size=8))
      print(p)
}

for (date_id in unique(mutation_df$date_stamp)){
  # subset df by variable
  mutation_sub=subset(mutation_df,date_stamp==date_id)
  
  # determine the top mutation frequency in each lineage
  # subset by top mutations
  mutation_sub_final=data.frame()
  for (lineage_id in unique(mutation_sub$lineage)){
    mutation_sub2=subset(mutation_sub,lineage==lineage_id)
    
    value_cutoff=(unique(mutation_sub2[
      order(mutation_sub2$percent_mutation,decreasing = TRUE),]$percent_mutation))[1]

    mutation_sub2=subset(mutation_sub2,percent_mutation==value_cutoff)
    mutation_sub_final=rbind(mutation_sub_final,mutation_sub2)
  }

  # determine the number of lineages in each date
  # if there are more than four, plot in groups of four
  lineage_list=unique(mutation_sub_final$lineage)
  for (i in seq(1,length(lineage_list), by=6)){
    if ((i+5) < length(lineage_list)){
      sub_list=unique(mutation_sub_final$lineage)[i:(i+5)]
      mutation_sub=subset(mutation_sub_final,lineage %in% sub_list)
      plot_by_date(mutation_sub)
    } else{
      sub_list=unique(mutation_sub_final$lineage)[i:length(lineage_list)]
      mutation_sub=subset(mutation_sub_final,lineage %in% sub_list)
      plot_by_date(mutation_sub)
    }
  }
}
```

### Review top mutations per lineage per timepoint
Determined by calculating the percent of a particular mutation being found in all project samples. Then determining the highest percentage and filtering for this value. For example, after sub-setting for one lineage:

```{r echo=FALSE}
tmp_df=data.frame()
kable(mutation_df[c(1,5,7,200),c("mutation","date_stamp","percent_mutation")])
```

The highest mutation rate is 10.61%. All mutations with this rate would be included.

```{r echo=FALSE}
for (lineage_id in unique(mutation_df$lineage)){
  # subset df by variable
  mutation_sub=subset(mutation_df,lineage==lineage_id)
  
  # subset by top mutations
  value_cutoff=median(unique(mutation_sub[
    order(mutation_sub$percent_mutation,decreasing = TRUE),]$percent_mutation))
  mutaton_sub2=subset(mutation_sub,percent_mutation >= value_cutoff)
  
  #Highlight mutation frequency
  print(paste0("Median mutation rate, across all projects, for lineage ", 
               lineage_id, ", is: ",value_cutoff,"%"))
  
  p = ggplot(data=mutaton_sub2,aes(x=mutation,y=percent_mutation )) +
    facet_wrap(~date_stamp,scales = "free_x",ncol=3) + 
    geom_bar(stat="identity", position=position_dodge()) +
    ggtitle(paste0("Percecnt of Samples with mutation by lineage: "),lineage_id) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text = element_text(size=8))
  print(p)
}

```

### Specific Mutation Analysis
Determine frequency over time of given mutation
```{r, echo=FALSE}
plot_mutations<-function(df_in){
    p = ggplot(data=df_in,aes(x=date_stamp,y=percent_mutation )) +
      facet_wrap(~lineage,scales = "free_x",ncol=3) + 
      geom_bar(stat="identity", position=position_dodge()) +
      ggtitle(paste0("Percecnt of Samples with mutation for mutation: "),mutation_id) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text = element_text(size=8))
    print(p)
}
for (mutation_id in mutation_list){
  mutation_sub_final=subset(mutation_df,mutation==mutation_id)
  
  if (nrow(mutation_sub_final)==0){
    print (paste0("Analyzed ", mutation_id, ": mutation not found"))
  } else{
    lineage_list=unique(mutation_sub_final$lineage)
    for (i in seq(1,length(lineage_list), by=6)){
      if ((i+5) < length(lineage_list)){
        sub_list=unique(mutation_sub_final$lineage)[i:(i+5)]
        mutation_sub=subset(mutation_sub_final,lineage %in% sub_list)
        plot_mutations(mutation_sub)
      } else{
        sub_list=unique(mutation_sub_final$lineage)[i:length(lineage_list)]
        mutation_sub=subset(mutation_sub_final,lineage %in% sub_list)
        plot_mutations(mutation_sub)
      }
    }
  }
}
```

